File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/ml/face/FaceImpl.kt
package dev.hebei.camerax_android.ml.face

import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceContour
import com.google.mlkit.vision.face.FaceLandmark
import dev.hebei.camerax_android.CameraXImpl
import dev.hebei.camerax_android.FaceContourTypeApi
import dev.hebei.camerax_android.FaceLandmarkTypeApi
import dev.hebei.camerax_android.PigeonApiFaceApi

class FaceImpl(impl: CameraXImpl) : PigeonApiFaceApi(impl) {
    override fun boundingBox(pigeon_instance: Face): android.graphics.Rect {
        return pigeon_instance.boundingBox
    }

    override fun allContours(pigeon_instance: Face): List<FaceContour> {
        return pigeon_instance.allContours
    }

    override fun allLandmarks(pigeon_instance: Face): List<FaceLandmark> {
        return pigeon_instance.allLandmarks
    }

    override fun getContour(pigeon_instance: Face, contourType: FaceContourTypeApi): FaceContour? {
        return pigeon_instance.getContour(contourType.impl)
    }

    override fun getLandmark(pigeon_instance: Face, landmarkType: FaceLandmarkTypeApi): FaceLandmark? {
        return pigeon_instance.getLandmark(landmarkType.impl)
    }

    override fun headEulerAngleX(pigeon_instance: Face): Double {
        return pigeon_instance.headEulerAngleX.toDouble()
    }

    override fun headEulerAngleY(pigeon_instance: Face): Double {
        return pigeon_instance.headEulerAngleY.toDouble()
    }

    override fun headEulerAngleZ(pigeon_instance: Face): Double {
        return pigeon_instance.headEulerAngleZ.toDouble()
    }

    override fun leftEyeOpenProbability(pigeon_instance: Face): Double? {
        return pigeon_instance.leftEyeOpenProbability?.toDouble()
    }

    override fun rightEyeOpenProbability(pigeon_instance: Face): Double? {
        return pigeon_instance.rightEyeOpenProbability?.toDouble()
    }

    override fun smilingProbability(pigeon_instance: Face): Double? {
        return pigeon_instance.smilingProbability?.toDouble()
    }

    override fun trackingId(pigeon_instance: Face): Long? {
        return pigeon_instance.trackingId?.toLong()
    }
}

File: camerax_android/lib/src/ml/face/face_impl.dart
import 'package:camerax_android/src/ml.dart';
import 'package:camerax_android/src/visionx.dart';
import 'package:camerax_android/src/api.dart';
import 'package:camerax_android/src/common.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class FaceImpl implements Face {
  final FaceProxyApi api;

  FaceImpl.internal(this.api);

  @override
  List<FaceContour> get allContours =>
      api.allContours.map((e) => e.impl).toList();
  @override
  List<FaceLandmark> get allLandmarks =>
      api.allLandmarks.map((e) => e.impl).toList();
  @override
  Rect<int> get boundingBox => api.boundingBox.impl;
  @override
  double get headEulerAngleX => api.headEulerAngleX;
  @override
  double get headEulerAngleY => api.headEulerAngleY;
  @override
  double get headEulerAngleZ => api.headEulerAngleZ;
  @override
  double? get leftEyeOpenProbability => api.leftEyeOpenProbability;
  @override
  double? get rightEyeOpenProbability => api.rightEyeOpenProbability;
  @override
  double? get smilingProbability => api.smilingProbability;
  @override
  int? get trackingId => api.trackingId;

  @override
  Future<FaceContour?> getContour(FaceContour$Type contourType) =>
      api.getContour(contourType.api).then((e) => e?.impl);

  @override
  Future<FaceLandmark?> getLandmark(FaceLandmark$Type landmarkType) =>
      api.getLandmark(landmarkType.api).then((e) => e?.impl);
}

extension FaceX on Face {
  VisionFaceObject visionFaceObject({required Size<int>? imageSize}) =>
      VisionFaceObjectImpl.internal(this, imageSize: imageSize);
}

extension FaceProxyApiX on FaceProxyApi {
  Face get impl => FaceImpl.internal(this);
}
